### PR TITLE
Don't convert Entity data if instance of ArrayAccess

### PR DESCRIPTION
**Pull request by othercorey**
Fixes https://github.com/cakephp/cakephp/issues/14482

Issues:

- Event::setData() will operate on the ArrayAccess instance

  Should we create a copy?

- IterateAggregate might provide a valid array conversion. 

  Not sure if we care about ArrayObject.

### CODE REVIEW

**Comment by dereuromark** [comment]
If we want to make this change without a clone, we should better put this into 4.next as this changes behavior now due to the mutable character of the input as you pointed out.

**Comment by ADmad** [comment]
> and a recursive cast here

PHP won't do a recursive cast.

**Comment by ADmad** [comment]
This problem has arisen because we assumed/forced event data to be an array in the past and have `getData()`, `setData()` which allow getting/setting individual keys.

We missed the opportunity in 4.0 to do a clean break and just store and return whatever is passed to the constructor without forcing/assuming it's type.

**Comment by markstory** [comment]
Event data has historically been expected to be an associative map of keys and values. For the linked issue the author could do `dispatch('event.name', $this, ['entity' => $thing])` and not have an issue.

**Comment by othercorey** [comment]
> Event data has historically been expected to be an associative map of keys and values. For the linked issue the author could do `dispatch('event.name', $this, ['entity' => $thing])` and not have an issue.

True. Perhaps instead of this work-around, we should have changed Event to enforce the array requirement so they don't end up with mangled data.

**Comment by othercorey** [comment]
I pushed an alternative fix to discuss.

- Changed default event data to `[]`
- Require event data to be an array or ArrayAccess. No casting.
- Throw exceptions when invalid event data is set.

File: src/Event/Event.php [hunk 1]
```diff
@@ -172,22 +179,26 @@ public function getData(?string $key = null)
...
      * @param mixed $value The value to set.
      * @return $this
+     * @throws \InvalidArgumentException When $key is not a string, array or does not implement \ArrayAccess.
      */
     public function setData($key, $value = null)
     {
-        if (is_array($key)) {
-            $this->_data = $key;
-        } else {
+        if (is_string($key)) {
```
Comment (ADmad) [inline src/Event/Event.php:195]: While it would be unwise one could use integer/numeric keys too.

File: src/Event/EventDispatcherInterface.php [hunk 2]
```diff
@@ -33,13 +33,13 @@ interface EventDispatcherInterface
...
      * @param string $name Name of the event.
-     * @param array|null $data Any value you wish to be transported with this event to
+     * @param array $data Any value you wish to be transported with this event to
      * it can be read by listeners.
      * @param object|null $subject The object that this event applies to
      * ($this by default).
      * @return \Cake\Event\EventInterface
      */
-    public function dispatchEvent(string $name, ?array $data = null, ?object $subject = null): EventInterface;
+    public function dispatchEvent(string $name, array $data = [], ?object $subject = null): EventInterface;
```
Comment (ADmad) [inline src/Event/EventDispatcherInterface.php:42]: Changing interface method signature would be technically a BC break.

File: src/Event/Event.php (hunk 1 above)
Comment (othercorey) [reply src/Event/Event.php:195]: True, but the $key typehint was already `string` so I used that.

File: src/Event/EventDispatcherInterface.php (hunk 2 above)
Comment (othercorey) [reply src/Event/EventDispatcherInterface.php:42]: Yea, but this shows what a better fix is. The default of null was a very odd choice which requires the array cast.

**Pull request closed without merging**